File: clients/php/src/Dialect.php
<?php
declare(strict_types=1);

namespace Orm;

/** The database-specific pieces of SQL for mysql, postgres, and sqlite. */
final class Dialect
{
    /** Replaced in trusted expression fragments with the advancing wall clock. */
    public const CURRENT_TIME_TOKEN = '$CURRENT_TIME';
    /** Sphere radius of MySQL ST_Distance_Sphere and the portable haversine. */
    private const EARTH_RADIUS = '6370986';

    /** Column types each column function accepts. */
    public const COLUMN_FUNCTION_TYPES = [
        'day_of_week' => ['date', 'datetime'],
        'year' => ['date', 'datetime'],
        'month' => ['date', 'datetime'],
        'date' => ['date', 'datetime'],
        'distance' => ['point'],
        'point_x' => ['point'],
        'point_y' => ['point'],
    ];

    /** Arguments of each column function, not counting the compared value. */
    public const COLUMN_FUNCTION_ARITY = ['day_of_week' => 0, 'year' => 0, 'month' => 0, 'date' => 0, 'distance' => 2, 'point_x' => 0, 'point_y' => 0];

    /** Interval unit of each relative value function. */
    public const VALUE_FUNCTION_UNITS = [
        'seconds_ago' => 'second', 'minutes_ago' => 'minute', 'hours_ago' => 'hour', 'days_ago' => 'day', 'months_ago' => 'month',
        'seconds_later' => 'second', 'minutes_later' => 'minute', 'hours_later' => 'hour', 'days_later' => 'day', 'months_later' => 'month',
    ];

    public function __construct(public readonly string $name)
    {
        if (!in_array($name, ['mysql', 'postgres', 'sqlite'], true)) {
            throw new OrmException(Code::DIALECT_UNKNOWN, $name);
        }
    }

    public static function isValueFunction(string $name): bool
    {
        return isset(self::VALUE_FUNCTION_UNITS[$name]) || $name === 'now' || $name === 'today';
    }

    public static function quoteWith(string $q, string $ident): string
    {
        return implode('.', array_map(static fn(string $p): string => $q . str_replace($q, $q . $q, $p) . $q, explode('.', $ident)));
    }

    public function quote(string $ident): string
    {
        return match ($this->name) {
            'mysql' => self::quoteWith('`', $ident),
            'postgres' => self::quoteWith('"', $ident),
            default => self::quoteWith('"', str_replace('.', '__', $ident)),
        };
    }

    public function placeholder(int $n): string
    {
        return $this->name === 'postgres' ? '$' . $n : '?';
    }

    public function limit(int $offset, int $count): string
    {
        return $this->name === 'mysql' ? " LIMIT $offset, $count" : " LIMIT $count OFFSET $offset";
    }

    public function forceIndex(string $index): string
    {
        return match ($this->name) {
            'mysql' => ' FORCE INDEX (' . self::quoteWith('`', $index) . ')',
            'postgres' => '',
            default => ' INDEXED BY ' . self::quoteWith('"', $index),
        };
    }

    public function insertReturningId(): bool
    {
        return $this->name !== 'mysql';
    }

    public function now(): string
    {
        return 'CURRENT_TIMESTAMP';
    }

    public function currentTime(): string
    {
        return $this->name === 'postgres' ? 'clock_timestamp()' : 'CURRENT_TIMESTAMP';
    }

    public function supports(string $op): bool
    {
        return $this->name !== 'sqlite' || ($op !== 'match' && $op !== 'match_boolean');
    }

    /** Whether a style stage is applied in SQL; the rest is left to the executor. */
    public function handlesStyle(string $style): bool
    {
        return match ($this->name) {
            'mysql' => $style === 'hex' || $style === 'ip',
            'postgres' => $style === 'ip',
            default => false,
        };
    }

    /** SQLite has no sub-second clock function; the executor binds the time. */
    public function hostNow(): bool
    {
        return $this->name === 'sqlite';
    }

    public function contains(string $col, string $ph): string
    {
        return match ($this->name) {
            'mysql' => "$col LIKE $ph",
            'postgres' => "$col ILIKE $ph",
            default => "$col LIKE $ph ESCAPE '\\'",
        };
    }

    /** @param \Closure(string): string $value a placeholder whose value gets the transform */
    public function containsBinary(string $col, \Closure $value): string
    {
        return match ($this->name) {
            'mysql' => "$col LIKE BINARY " . $value('like_contains'),
            'postgres' => "$col LIKE " . $value('like_contains'),
            default => "instr($col, " . $value('') . ') > 0',
        };
    }

    /** @param list<string> $cols */
    public function fulltext(array $cols, string $ph, bool $boolean): string
    {
        if ($this->name === 'mysql') {
            return 'MATCH(' . implode(', ', $cols) . ") AGAINST ($ph" . ($boolean ? ' IN BOOLEAN MODE' : ' IN NATURAL LANGUAGE MODE') . ')';
        }
        $doc = count($cols) > 1 ? 'coalesce(' . implode(", '') || ' ' || coalesce(", $cols) . ", '')" : implode(" || ' ' || ", $cols);
        $fn = $boolean ? 'websearch_to_tsquery' : 'plainto_tsquery';
        return "to_tsvector('simple', $doc) @@ $fn('simple', $ph)";
    }

    /** @param list<string> $conflict */
    public function upsert(array $conflict, string $assigns): string
    {
        if ($this->name === 'mysql') {
            return ' ON DUPLICATE KEY UPDATE ' . $assigns;
        }
        $q = array_map(static fn(string $c): string => self::quoteWith('"', $c), $conflict);
        return ' ON CONFLICT (' . implode(', ', $q) . ') DO UPDATE SET ' . $assigns;
    }

    /** SQL-side read stages of a column. @param list<string> $styles */
    public function readExpr(string $col, string $type, array $styles): string
    {
        switch ($this->name) {
            case 'mysql':
                if ($type === 'point') {
                    $col = "ST_AsText($col)";
                }
                for ($i = count($styles) - 1; $i >= 0; $i--) {
                    $col = match ($styles[$i]) {
                        'hex' => "UNHEX($col)",
                        'ip' => "INET6_NTOA($col)",
                        default => $col,
                    };
                }
                return $col;
            case 'postgres':
                if ($type === 'point') {
                    $col = "($col)::text";
                }
                return in_array('ip', $styles, true) ? "host($col)" : $col;
            default:
                return $col;
        }
    }

    /** SQL-side write stages around a bound value. @param list<string> $styles */
    public function writeExpr(string $ph, string $type, array $styles): string
    {
        switch ($this->name) {
            case 'mysql':
                if ($type === 'point') {
                    $ph = "ST_PointFromText($ph)";
                }
                foreach ($styles as $s) {
                    $ph = match ($s) {
                        'hex' => "HEX($ph)",
                        'ip' => "INET6_ATON($ph)",
                        default => $ph,
                    };
                }
                return $ph;
            case 'postgres':
                if ($type === 'point') {
                    $ph = "CAST($ph AS text)::point";
                }
                foreach ($styles as $s) {
                    if ($s === 'ip') {
                        $ph = "($ph)::inet";
                    }
                }
                return $ph;
            default:
                return $ph;
        }
    }

    /** The row lock suffix; SQLite locks in the executor. */
    public function rowLock(string $mode): ?string
    {
        if (!in_array($mode, ['update', 'share', 'update_nowait', 'share_nowait'], true)) {
            return null;
        }
        if ($this->name === 'sqlite') {
            return '';
        }
        return ' FOR ' . strtoupper(str_replace('_', ' ', $mode));
    }

    public function random(): string
    {
        return $this->name === 'mysql' ? 'RAND()' : 'random()';
    }

    /**
     * @param list<string> $cols
     * @param list<list<string>> $rows
     */
    public function tupleIn(array $cols, array $rows, bool $negate): string
    {
        $list = implode(', ', array_map(static fn(array $r): string => '(' . implode(', ', $r) . ')', $rows));
        if ($this->name === 'sqlite') {
            $list = 'VALUES ' . $list;
        }
        return '(' . implode(', ', $cols) . ')' . ($negate ? ' NOT IN ' : ' IN ') . '(' . $list . ')';
    }

    /** @param \Closure(int): string $arg a placeholder for the i-th function argument */
    public function columnFunction(string $fn, string $col, \Closure $arg): ?string
    {
        switch ($this->name) {
            case 'mysql':
                return match ($fn) {
                    'day_of_week' => "DAYOFWEEK($col)",
                    'year' => "YEAR($col)",
                    'month' => "MONTH($col)",
                    'date' => "DATE($col)",
                    'distance' => "ST_Distance_Sphere($col, POINT(" . $arg(0) . ', ' . $arg(1) . '))',
                    'point_x' => "ST_X($col)",
                    'point_y' => "ST_Y($col)",
                    default => null,
                };
            case 'postgres':
                return match ($fn) {
                    'day_of_week' => "(EXTRACT(DOW FROM $col)::int + 1)",
                    'year' => "EXTRACT(YEAR FROM $col)::int",
                    'month' => "EXTRACT(MONTH FROM $col)::int",
                    'date' => "CAST($col AS date)",
                    'distance' => self::haversine("{$col}[0]", "{$col}[1]", static fn(): string => 'CAST(' . $arg(0) . ' AS double precision)', static fn(): string => 'CAST(' . $arg(1) . ' AS double precision)'),
                    'point_x' => "{$col}[0]",
                    'point_y' => "{$col}[1]",
                    default => null,
                };
            default:
                return match ($fn) {
                    'day_of_week' => "(CAST(strftime('%w', $col) AS INTEGER) + 1)",
                    'year' => "CAST(strftime('%Y', $col) AS INTEGER)",
                    'month' => "CAST(strftime('%m', $col) AS INTEGER)",
                    'date' => "date($col)",
                    'distance' => self::haversine(self::sqlitePoint($col, false), self::sqlitePoint($col, true), static fn(): string => $arg(0), static fn(): string => $arg(1)),
                    'point_x' => self::sqlitePoint($col, false),
                    'point_y' => self::sqlitePoint($col, true),
                    default => null,
                };
        }
    }

    /**
     * @param \Closure(): string $arg the interval amount placeholder
     * @param \Closure(): string $now the executor clock placeholder
     */
    public function valueFunction(string $fn, \Closure $arg, \Closure $now): ?string
    {
        $unit = self::VALUE_FUNCTION_UNITS[$fn] ?? null;
        $later = str_ends_with($fn, '_later');
        switch ($this->name) {
            case 'mysql':
                return match (true) {
                    $fn === 'now' => 'NOW()',
                    $fn === 'today' => 'CURDATE()',
                    $unit === null => null,
                    default => ($later ? 'DATE_ADD' : 'DATE_SUB') . '(NOW(), INTERVAL ' . $arg() . ' ' . strtoupper($unit) . ')',
                };
            case 'postgres':
                if ($fn === 'now') {
                    return 'now()';
                }
                if ($fn === 'today') {
                    return 'CURRENT_DATE';
                }
                if ($unit === null) {
                    return null;
                }
                $field = ['second' => 'secs', 'minute' => 'mins', 'hour' => 'hours', 'day' => 'days', 'month' => 'months'][$unit];
                $cast = $field === 'secs' ? 'double precision' : 'integer';
                $value = 'CAST(' . $arg() . " AS $cast)";
                return '(now()' . ($later ? ' + ' : ' - ') . "make_interval($field => $value))";
            default:
                if ($fn === 'now') {
                    return $now();
                }
                if ($fn === 'today') {
                    return 'date(' . $now() . ')';
                }
                if ($unit === null) {
                    return null;
                }
                $clock = $now();
                $modifier = ($later ? "'+'" : "'-'") . ' || CAST(' . $arg() . " AS TEXT) || ' {$unit}s'";
                return $unit === 'month' ? "datetime($clock, $modifier, 'floor')" : "datetime($clock, $modifier)";
        }
    }

    /** Great-circle distance; $x2 and $y2 bind a value at each occurrence. */
    private static function haversine(string $x1, string $y1, \Closure $x2, \Closure $y2): string
    {
        $rad = static fn(string $v): string => "RADIANS($v)";
        return '(2 * ' . self::EARTH_RADIUS . ' * ASIN(SQRT(POWER(SIN((' . $rad($y2()) . ' - ' . $rad($y1) . ') / 2), 2) + COS(' . $rad($y1) . ') * COS(' . $rad($y2()) . ') * POWER(SIN((' . $rad($x2()) . ' - ' . $rad($x1) . ') / 2), 2))))';
    }

    /** One coordinate of the stored `POINT(x y)` text. */
    private static function sqlitePoint(string $col, bool $second): string
    {
        $space = "instr($col, ' ')";
        if ($second) {
            return "CAST(substr($col, $space + 1, length($col) - $space - 1) AS REAL)";
        }
        return "CAST(substr($col, 7, $space - 7) AS REAL)";
    }
}

File: clients/rust/orm/src/engine/dialect.rs
//! The database-specific pieces of SQL. The planner never writes a quote or a
//! placeholder itself.

/// Replaced in code-owned expression fragments with the dialect's advancing
/// wall-clock expression.
pub const CURRENT_TIME_TOKEN: &str = "$CURRENT_TIME";

/// The sphere radius of MySQL `ST_Distance_Sphere`, used by the portable
/// haversine rendering.
const EARTH_RADIUS_METERS: &str = "6370986";

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Dialect {
    MySql,
    Postgres,
    Sqlite,
}

/// The column types each column function accepts.
pub fn column_function_types(name: &str) -> Option<&'static [&'static str]> {
    Some(match name {
        "day_of_week" | "year" | "month" | "date" => &["date", "datetime"],
        "distance" | "point_x" | "point_y" => &["point"],
        _ => return None,
    })
}

/// The function arguments of a column function, not counting the compared value.
pub fn column_function_arity(name: &str) -> usize {
    if name == "distance" {
        2
    } else {
        0
    }
}

/// The interval unit of a relative value function.
pub fn value_function_unit(name: &str) -> Option<&'static str> {
    Some(match name {
        "seconds_ago" | "seconds_later" => "second",
        "minutes_ago" | "minutes_later" => "minute",
        "hours_ago" | "hours_later" => "hour",
        "days_ago" | "days_later" => "day",
        "months_ago" | "months_later" => "month",
        _ => return None,
    })
}

pub fn is_value_function(name: &str) -> bool {
    value_function_unit(name).is_some() || name == "now" || name == "today"
}

fn quote_with(q: &str, ident: &str) -> String {
    ident.split('.').map(|part| format!("{q}{}{q}", part.replace(q, &format!("{q}{q}")))).collect::<Vec<_>>().join(".")
}

/// The great-circle distance. `x2` and `y2` are called once per occurrence so
/// positional placeholders receive one bind each.
fn haversine(x1: &str, y1: &str, x2: &mut dyn FnMut() -> String, y2: &mut dyn FnMut() -> String) -> String {
    let rad = |v: String| format!("RADIANS({v})");
    let a = rad(y2());
    let b = rad(y2());
    let c = rad(x2());
    format!(
        "(2 * {EARTH_RADIUS_METERS} * ASIN(SQRT(POWER(SIN(({a} - {y1r}) / 2), 2) + COS({y1r}) * COS({b}) * POWER(SIN(({c} - {x1r}) / 2), 2))))",
        y1r = rad(y1.to_owned()),
        x1r = rad(x1.to_owned()),
    )
}

fn sqlite_point_coordinate(col: &str, second: bool) -> String {
    let space = format!("instr({col}, ' ')");
    if second {
        format!("CAST(substr({col}, {space} + 1, length({col}) - {space} - 1) AS REAL)")
    } else {
        format!("CAST(substr({col}, 7, {space} - 7) AS REAL)")
    }
}

fn tuple_in(cols: &[String], rows: &[Vec<String>], negate: bool, values: bool) -> String {
    let parts: Vec<String> = rows.iter().map(|r| format!("({})", r.join(", "))).collect();
    let op = if negate { " NOT IN " } else { " IN " };
    let mut list = parts.join(", ");
    if values {
        list = format!("VALUES {list}");
    }
    format!("({}){op}({list})", cols.join(", "))
}

impl Dialect {
    pub fn parse(name: &str) -> Option<Dialect> {
        match name {
            "mysql" => Some(Dialect::MySql),
            "postgres" => Some(Dialect::Postgres),
            "sqlite" => Some(Dialect::Sqlite),
            _ => None,
        }
    }

    pub fn name(self) -> &'static str {
        match self {
            Dialect::MySql => "mysql",
            Dialect::Postgres => "postgres",
            Dialect::Sqlite => "sqlite",
        }
    }

    pub fn quote(self, ident: &str) -> String {
        match self {
            Dialect::MySql => quote_with("`", ident),
            Dialect::Postgres => quote_with("\"", ident),
            Dialect::Sqlite => quote_with("\"", &ident.split('.').collect::<Vec<_>>().join("__")),
        }
    }

    /// The placeholder of the n-th (1-based) bind.
    pub fn placeholder(self, n: usize) -> String {
        match self {
            Dialect::Postgres => format!("${n}"),
            _ => "?".into(),
        }
    }

    pub fn limit(self, offset: u32, count: u32) -> String {
        match self {
            Dialect::MySql => format!(" LIMIT {offset}, {count}"),
            _ => format!(" LIMIT {count} OFFSET {offset}"),
        }
    }

    pub fn force_index(self, name: &str) -> String {
        match self {
            Dialect::MySql => format!(" FORCE INDEX ({})", quote_with("`", name)),
            Dialect::Postgres => String::new(),
            Dialect::Sqlite => format!(" INDEXED BY {}", quote_with("\"", name)),
        }
    }

    pub fn insert_returning_id(self) -> bool {
        self != Dialect::MySql
    }

    pub fn now(self) -> &'static str {
        "CURRENT_TIMESTAMP"
    }

    pub fn current_time(self) -> &'static str {
        match self {
            Dialect::Postgres => "clock_timestamp()",
            _ => "CURRENT_TIMESTAMP",
        }
    }

    /// Whether a predicate operator exists in this dialect.
    pub fn supports(self, op: &str) -> bool {
        self != Dialect::Sqlite || (op != "match" && op != "match_boolean")
    }

    /// Whether a column style stage is applied in SQL.
    pub fn handles_style(self, style: &str) -> bool {
        match self {
            Dialect::MySql => style == "hex" || style == "ip",
            Dialect::Postgres => style == "ip",
            Dialect::Sqlite => false,
        }
    }

    /// Whether ORM-written timestamps come from the executor clock.
    pub fn host_now(self) -> bool {
        self == Dialect::Sqlite
    }

    /// The row lock suffix; SQLite locks through the executor.
    pub fn row_lock(self, mode: &str) -> Option<&'static str> {
        if !matches!(mode, "update" | "share" | "update_nowait" | "share_nowait") {
            return None;
        }
        if self == Dialect::Sqlite {
            return Some("");
        }
        Some(match mode {
            "update" => " FOR UPDATE",
            "share" => " FOR SHARE",
            "update_nowait" => " FOR UPDATE NOWAIT",
            _ => " FOR SHARE NOWAIT",
        })
    }

    pub fn like(self, col: &str, ph: &str) -> String {
        match self {
            Dialect::MySql => format!("{col} LIKE {ph}"),
            Dialect::Postgres => format!("{col} ILIKE {ph}"),
            Dialect::Sqlite => format!("{col} LIKE {ph} ESCAPE '\\'"),
        }
    }

    pub fn contains_binary(self, col: &str, value: &mut dyn FnMut(&str) -> String) -> String {
        match self {
            Dialect::MySql => format!("{col} LIKE BINARY {}", value("like_contains")),
            Dialect::Postgres => format!("{col} LIKE {}", value("like_contains")),
            Dialect::Sqlite => format!("instr({col}, {}) > 0", value("")),
        }
    }

    pub fn fulltext(self, cols: &[String], ph: &str, boolean: bool) -> String {
        match self {
            Dialect::MySql => {
                let mode = if boolean { " IN BOOLEAN MODE" } else { " IN NATURAL LANGUAGE MODE" };
                format!("MATCH({}) AGAINST ({ph}{mode})", cols.join(", "))
            }
            _ => {
                let doc = if cols.len() > 1 { format!("coalesce({}, '')", cols.join(", '') || ' ' || coalesce(")) } else { cols.join(" || ' ' || ") };
                let f = if boolean { "websearch_to_tsquery" } else { "plainto_tsquery" };
                format!("to_tsvector('simple', {doc}) @@ {f}('simple', {ph})")
            }
        }
    }

    pub fn upsert(self, conflict: &[String], assigns: &str) -> String {
        match self {
            Dialect::MySql => format!(" ON DUPLICATE KEY UPDATE {assigns}"),
            _ => {
                let q: Vec<String> = conflict.iter().map(|c| quote_with("\"", c)).collect();
                format!(" ON CONFLICT ({}) DO UPDATE SET {assigns}", q.join(", "))
            }
        }
    }

    /// Wraps the SQL-side read stages of a column.
    pub fn read_expr(self, col: &str, col_type: &str, styles: &[&str]) -> String {
        match self {
            Dialect::MySql => {
                let mut expr = if col_type == "point" { format!("ST_AsText({col})") } else { col.to_owned() };
                for s in styles.iter().rev() {
                    match *s {
                        "hex" => expr = format!("UNHEX({expr})"),
                        "ip" => expr = format!("INET6_NTOA({expr})"),
                        _ => {}
                    }
                }
                expr
            }
            Dialect::Postgres => {
                let col = if col_type == "point" { format!("({col})::text") } else { col.to_owned() };
                if styles.contains(&"ip") {
                    return format!("host({col})");
                }
                col
            }
            Dialect::Sqlite => col.to_owned(),
        }
    }

    /// Wraps a bound value with the SQL-side write stages of a column.
    pub fn write_expr(self, ph: String, col_type: &str, styles: &[&str]) -> String {
        match self {
            Dialect::MySql => {
                let mut expr = if col_type == "point" { format!("ST_PointFromText({ph})") } else { ph };
                for s in styles {
                    match *s {
                        "hex" => expr = format!("HEX({expr})"),
                        "ip" => expr = format!("INET6_ATON({expr})"),
                        _ => {}
                    }
                }
                expr
            }
            Dialect::Postgres => {
                let mut expr = if col_type == "point" { format!("CAST({ph} AS text)::point") } else { ph };
                for s in styles {
                    if *s == "ip" {
                        expr = format!("({expr})::inet");
                    }
                }
                expr
            }
            Dialect::Sqlite => ph,
        }
    }

    /// Renders a column function applied to `col`; `arg(i)` binds the i-th
    /// function argument.
    pub fn column_function(self, name: &str, col: &str, arg: &mut dyn FnMut(usize) -> String) -> Option<String> {
        Some(match (self, name) {
            (Dialect::MySql, "day_of_week") => format!("DAYOFWEEK({col})"),
            (Dialect::MySql, "year") => format!("YEAR({col})"),
            (Dialect::MySql, "month") => format!("MONTH({col})"),
            (Dialect::MySql, "date") => format!("DATE({col})"),
            (Dialect::MySql, "distance") => {
                let (x, y) = (arg(0), arg(1));
                format!("ST_Distance_Sphere({col}, POINT({x}, {y}))")
            }
            (Dialect::MySql, "point_x") => format!("ST_X({col})"),
            (Dialect::MySql, "point_y") => format!("ST_Y({col})"),
            (Dialect::Postgres, "day_of_week") => format!("(EXTRACT(DOW FROM {col})::int + 1)"),
            (Dialect::Postgres, "year") => format!("EXTRACT(YEAR FROM {col})::int"),
            (Dialect::Postgres, "month") => format!("EXTRACT(MONTH FROM {col})::int"),
            (Dialect::Postgres, "date") => format!("CAST({col} AS date)"),
            (Dialect::Postgres, "distance") => {
                let cell = std::cell::RefCell::new(arg);
                let mut x = || format!("CAST({} AS double precision)", (cell.borrow_mut())(0));
                let mut y = || format!("CAST({} AS double precision)", (cell.borrow_mut())(1));
                haversine(&format!("{col}[0]"), &format!("{col}[1]"), &mut x, &mut y)
            }
            (Dialect::Postgres, "point_x") => format!("{col}[0]"),
            (Dialect::Postgres, "point_y") => format!("{col}[1]"),
            (Dialect::Sqlite, "day_of_week") => format!("(CAST(strftime('%w', {col}) AS INTEGER) + 1)"),
            (Dialect::Sqlite, "year") => format!("CAST(strftime('%Y', {col}) AS INTEGER)"),
            (Dialect::Sqlite, "month") => format!("CAST(strftime('%m', {col}) AS INTEGER)"),
            (Dialect::Sqlite, "date") => format!("date({col})"),
            (Dialect::Sqlite, "distance") => {
                let cell = std::cell::RefCell::new(arg);
                let mut x = || (cell.borrow_mut())(0);
                let mut y = || (cell.borrow_mut())(1);
                haversine(&sqlite_point_coordinate(col, false), &sqlite_point_coordinate(col, true), &mut x, &mut y)
            }
            (Dialect::Sqlite, "point_x") => sqlite_point_coordinate(col, false),
            (Dialect::Sqlite, "point_y") => sqlite_point_coordinate(col, true),
            _ => return None,
        })
    }

    /// Renders a value function; `arg` binds the interval amount and `now`
    /// binds the executor clock, both in SQL text order.
    pub fn value_function(self, name: &str, arg: &mut dyn FnMut() -> String, now: &mut dyn FnMut() -> String) -> Option<String> {
        match (self, name) {
            (Dialect::MySql, "now") => return Some("NOW()".into()),
            (Dialect::MySql, "today") => return Some("CURDATE()".into()),
            (Dialect::Postgres, "now") => return Some("now()".into()),
            (Dialect::Postgres, "today") => return Some("CURRENT_DATE".into()),
            (Dialect::Sqlite, "now") => return Some(now()),
            (Dialect::Sqlite, "today") => return Some(format!("date({})", now())),
            _ => {}
        }
        let unit = value_function_unit(name)?;
        let later = name.ends_with("_later");
        Some(match self {
            Dialect::MySql => {
                let f = if later { "DATE_ADD" } else { "DATE_SUB" };
                format!("{f}(NOW(), INTERVAL {} {})", arg(), unit.to_uppercase())
            }
            Dialect::Postgres => {
                let op = if later { " + " } else { " - " };
                let field = match unit {
                    "second" => "secs",
                    "minute" => "mins",
                    "hour" => "hours",
                    "day" => "days",
                    _ => "months",
                };
                let cast = if field == "secs" { "double precision" } else { "integer" };
                format!("(now(){op}make_interval({field} => CAST({} AS {cast})))", arg())
            }
            Dialect::Sqlite => {
                let sign = if later { "'+'" } else { "'-'" };
                let clock = now();
                let modifier = format!("{sign} || CAST({} AS TEXT) || ' {unit}s'", arg());
                if unit == "month" {
                    format!("datetime({clock}, {modifier}, 'floor')")
                } else {
                    format!("datetime({clock}, {modifier})")
                }
            }
        })
    }

    pub fn tuple_in(self, cols: &[String], rows: &[Vec<String>], negate: bool) -> String {
        tuple_in(cols, rows, negate, self == Dialect::Sqlite)
    }

    pub fn random(self) -> &'static str {
        match self {
            Dialect::MySql => "RAND()",
            _ => "random()",
        }
    }
}
